test/extended/apiserver/webhook.go

```diff
@@ -0,0 +1,292 @@
+		},
+		Webhooks: []admissionregistrationv1.ValidatingWebhook{
+			{
+				Name: "deny-all-webhook.openshift.io",
+				Rules: []admissionregistrationv1.RuleWithOperations{{
+					Operations: []admissionregistrationv1.OperationType{admissionregistrationv1.Create},
+					Rule: admissionregistrationv1.Rule{
+						APIGroups:   []string{"*"},
+						APIVersions: []string{"*"},
+						Resources:   []string{"*"},
+					},
+				}},
+				ClientConfig: admissionregistrationv1.WebhookClientConfig{
+					Service: &admissionregistrationv1.ServiceReference{
+						Namespace: namespace,
+						Name:      serviceName,
+						Path:      strPtr("/always-deny"),
+						Port:      pointer.Int32Ptr(servicePort),
+					},
+				},
+				SideEffects:             &sideEffectsNone,
+				AdmissionReviewVersions: []string{"v1", "v1beta1"},
+				FailurePolicy:           &failurePolicy,
+				NamespaceSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{namespaceSelector: "true"},
+				},
+				ObjectSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{objectSelector: "true"},
+				},
+			},
+		},
+	}
+	_, err = client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Create(context.Background(), config, metav1.CreateOptions{})
+	e2e.ExpectNoError(err, "creating webhook config %s in namespace %s", configName, namespace)
+
+	removeWebhookConfiguration := func() {
+		err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.Background(), configName, metav1.DeleteOptions{})
+		e2e.ExpectNoError(err, "deleting webhook config %s in namespace %s", configName, namespace)
+	}
+
+	err = hitValidatingWebhook(func() (func(), error) {
+		cmClient := client.CoreV1().ConfigMaps(namespace)
+		cm := &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: string(uuid.NewUUID()),
+				Labels: map[string]string{
+					objectSelector: "true",
+				},
+			},
+		}
+		_, err := cmClient.Create(context.Background(), cm, metav1.CreateOptions{})
+		removeConfigMap := func() {
+			err = cmClient.Delete(context.Background(), cm.GetName(), metav1.DeleteOptions{})
+			e2e.ExpectNoError(err, "deleting config map %s in namespace %s", cm.Name, namespace)
+		}
+
+		return removeConfigMap, err
+	})
+	if err != nil {
+		removeWebhookConfiguration()
+		e2e.ExpectNoError(err, "calling validating admission webhook")
+	}
+	return removeWebhookConfiguration
+}
+
+// hitValidatingWebhook tries to create a resource and expect it to fail (deny)
+// it tries to send a few request before giving up because registering a webhook is not instant and might take up to a few seconds
+func hitValidatingWebhook(callback func() (func(), error)) error {
+	return wait.PollImmediate(100*time.Millisecond, 30*time.Second, func() (bool, error) {
+		cleanup, err := callback()
+		if err != nil {
+			// the always-deny webhook does not provide a reason, so check for the error string we expect
+			if strings.Contains(err.Error(), "denied") {
```

Comment:
what is the full error string?

Reply:
the full error is `admission webhook "deny-all-webhook.openshift.io" denied the request: this webhook denies all requests`